libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyManager.java

```diff
@@ -56,27 +84,120 @@ public PolicyManager(Policy defaultPolicy, Map<String, Policy> pluginPolicies, F
         this.pluginResolver = pluginResolver;
     }
 
-    public void checkFlagEntitlement(Class<?> callerClass, FlagEntitlementType type) {
+    private static List<Entitlement> lookupEntitlementsForModule(Policy policy, String moduleName) {
+        for (int i = 0; i < policy.scopes.size(); ++i) {
+            var scope = policy.scopes.get(i);
+            if (scope.name.equals(moduleName)) {
+                return scope.entitlements;
+            }
+        }
+        return null;
+    }
+
+    public void checkExitVMEntitlement(Class<?> callerClass) {
+        checkEntitlementPresent(callerClass, ExitVMEntitlement.class);
+    }
+
+    public void checkCreateClassLoaderEntitlement(Class<?> callerClass) {
+        checkEntitlementPresent(callerClass, CreateClassLoaderEntitlement.class);
+    }
+
+    private void checkEntitlementPresent(Class<?> callerClass, Class<? extends Entitlement> entitlementClass) {
         var requestingModule = requestingModule(callerClass);
         if (isTriviallyAllowed(requestingModule)) {
             return;
         }
 
-        // TODO: real policy check. For now, we only allow our hardcoded System.exit policy for server.
-        // TODO: this will be checked using policies
-        if (requestingModule.isNamed()
-            && requestingModule.getName().equals("org.elasticsearch.server")
-            && (type == FlagEntitlementType.SYSTEM_EXIT || type == FlagEntitlementType.CREATE_CLASSLOADER)) {
-            logger.debug("Allowed: caller [{}] in module [{}] has entitlement [{}]", callerClass, requestingModule.getName(), type);
+        ModuleEntitlements entitlements = getEntitlementsOrThrow(callerClass, requestingModule);
+        if (entitlements.hasEntitlement(entitlementClass)) {
+            logger.debug(
+                "Allowed: caller [{}] in module [{}] has [{}]",
+                callerClass,
+                requestingModule.getName(),
+                entitlementClass.getSimpleName()
+            );
             return;
         }
-
-        // TODO: plugins policy check using pluginResolver and pluginPolicies
         throw new NotEntitledException(
-            Strings.format("Missing entitlement [%s] for caller [%s] in module [%s]", type, callerClass, requestingModule.getName())
+            Strings.format(
+                "Caller [%s] in module [%s] does not have [%s]",
```

Comment:
Nit: We'll find these easier to read if the message is not interleaved with long variable-length strings. The syntax from before this change is probably easier to deal with in practice, because it starts right off with the headline `Missing entitlement` and then adds detail.

As a thought experiment, consider if we wanted to alert on this. Would we want to match the text `does not have`? I'd think `Missing entitlement` would be clearer and more selective.

This nit applies to most of the exception messages in this file.